Sources/Core/Shared/GroupOperation.swift

```diff
@@ -225,12 +305,41 @@ public class GroupOperation: Operation, OperationQueueDelegate {
     }
 
     public func operationQueue(queue: OperationQueue, didFinishOperation operation: NSOperation, withErrors errors: [ErrorType]) {
+        guard queue === self.queue else { return }
 
         if operation === finishingOperation {
             finish(fatalErrors)
             queue.suspended = true
         }
     }
+
+    public func operationQueue(queue: OperationQueue, willProduceOperation operation: NSOperation) {
+        guard queue === self.queue else { return }
+
+        // Ensure that produced operations are added to GroupOperation's
+        // internal operations array (and cancelled if appropriate)
+
+        let shouldContinue = groupFinishLock.withCriticalScope { () -> Bool in
+            assert(!finishingOperation.finished, "Cannot produce new operations within a group after the group has completed.")
+            guard !isGroupFinishing else {
+                assertionFailure("Cannot produce new operations within a group after the group has started to finish.")
+                return false
+            }
+            dispatch_group_enter(isAddingOperationsGroup)
+            return true
+        }
+
+        guard shouldContinue else { return }
+
+        _operations.append(operation)
+        if cancelled {
+            if !operation.cancelled { operation.cancel() }
+        }
```

Comment:
Can we re-write this as:

```swift
if cancelled && !operation.cancelled {
    operation.cancel()
}
```

It took me a little while to try to figure out if this was a trailing closure or not.